From 51af41173fa6f266b1c9649e91af618a3a96b0e5 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:09 -0400
Subject: [fs] xprtrdma: Disable completions for FAST_REG_MR Work Requests

Message-id: <1411050057-12417-47-git-send-email-steved@redhat.com>
Patchwork-id: 96207
O-Subject: [RHEL7.1 PATCH 46/94] xprtrdma: Disable completions for FAST_REG_MR Work Requests
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Chuck Lever <chuck.lever@oracle.com>

Instead of relying on a completion to change the state of an FRMR
to FRMR_IS_VALID, set it in advance. If an error occurs, a completion
will fire anyway and mark the FRMR FRMR_IS_STALE.

Signed-off-by: Chuck Lever <chuck.lever@oracle.com>
Tested-by: Steve Wise <swise@opengridcomputing.com>
Tested-by: Shirley Ma <shirley.ma@oracle.com>
Tested-by: Devesh Sharma <devesh.sharma@emulex.com>
Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>
Upstream Id: 050557220e34ed5acc830c9bf6cd993f6b4ea33e
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/xprtrdma/verbs.c b/net/sunrpc/xprtrdma/verbs.c
index 7459b86..3e8b388 100644
--- a/net/sunrpc/xprtrdma/verbs.c
+++ b/net/sunrpc/xprtrdma/verbs.c
@@ -159,10 +159,7 @@ rpcrdma_sendcq_process_wc(struct ib_wc *wc)
   return;
  }
 
- if (wc->opcode == IB_WC_FAST_REG_MR)
-  frmr->r.frmr.fr_state = FRMR_IS_VALID;
- else if (wc->opcode == IB_WC_LOCAL_INV)
-  frmr->r.frmr.fr_state = FRMR_IS_INVALID;
+ frmr->r.frmr.fr_state = FRMR_IS_INVALID;
 }
 
 static int
@@ -1727,10 +1724,11 @@ rpcrdma_register_frmr_external(struct rpcrdma_mr_seg *seg,
  dprintk("RPC:       %s: Using frmr %p to map %d segments\n",
   __func__, mw, i);
 
+ frmr->fr_state = FRMR_IS_VALID;
+
  memset(&frmr_wr, 0, sizeof frmr_wr);
  frmr_wr.wr_id = (unsigned long)(void *)mw;
  frmr_wr.opcode = IB_WR_FAST_REG_MR;
- frmr_wr.send_flags = IB_SEND_SIGNALED;
  frmr_wr.wr.fast_reg.iova_start = seg1->mr_dma;
  frmr_wr.wr.fast_reg.page_list = frmr->fr_pgl;
  frmr_wr.wr.fast_reg.page_list_len = page_no;
@@ -1766,6 +1764,7 @@ rpcrdma_register_frmr_external(struct rpcrdma_mr_seg *seg,
  *nsegs = i;
  return 0;
 out_err:
+ frmr->fr_state = FRMR_IS_INVALID;
  while (i--)
   rpcrdma_unmap_one(ia, --seg);
  return rc;
-- 
1.7.1